tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/BoundaryConditions.hpp

```diff
@@ -0,0 +1,614 @@
+      make_with_random_values<
+          tnsr::i<DataVector, FaceDim + 1, Frame::Inertial>>(
+          generator, make_not_null(&dist), number_of_points_on_face);
+
+  if constexpr (has_inv_spatial_metric) {
+    auto& inv_spatial_metric =
+        get<tmpl::front<inverse_spatial_metric_list>>(interior_face_fields);
+    detail::adjust_inverse_spatial_metric(make_not_null(&inv_spatial_metric));
+    auto& normal_vector =
+        get<::evolution::dg::Actions::detail::NormalVector<FaceDim + 1>>(
+            interior_face_fields);
+    detail::normalize_vector_and_covector(
+        make_not_null(&interior_normal_covector), make_not_null(&normal_vector),
+        inv_spatial_metric);
+  } else {
+    const Scalar<DataVector> magnitude = ::magnitude(interior_normal_covector);
+    for (DataVector& component : interior_normal_covector) {
+      component /= get(magnitude);
+    }
+  }
+
+  // Set a random mesh velocity. We allow for velocities above 1 to test "faster
+  // than light" mesh movement.
+  std::optional<tnsr::I<DataVector, FaceDim + 1>> face_mesh_velocity{};
+  if (use_moving_mesh) {
+    std::uniform_real_distribution<> local_dist(-2., 2.);
+    face_mesh_velocity =
+        make_with_random_values<tnsr::I<DataVector, FaceDim + 1>>(
+            generator, make_not_null(&local_dist), number_of_points_on_face);
+  }
+
+  if constexpr (BoundaryCondition::bc_type ==
+                ::evolution::BoundaryConditions::Type::Outflow) {
+    // Outflow boundary conditions only check that all characteristic speeds
+    // are directed out of the element. If there are any inward directed
+    // fields then the boundary condition should error.
+    const auto apply_bc =
+        [&boundary_condition, &face_mesh_velocity, &interior_normal_covector,
+         &python_boundary_condition_functions,
+         &python_module](const auto&... face_and_volume_args) noexcept {
+          const std::optional<std::string> error_msg =
+              boundary_condition.dg_outflow(face_mesh_velocity,
+                                            interior_normal_covector,
+                                            face_and_volume_args...);
+          const std::string& python_error_msg_function =
+              get_python_error_message_function<BoundaryCorrection>(
+                  python_boundary_condition_functions);
+          const auto python_error_message =
+              call_for_error_message<ConversionClassList>(
+                  python_module, python_error_msg_function, face_mesh_velocity,
+                  interior_normal_covector, face_and_volume_args...);
+          CHECK(python_error_message.has_value() == error_msg.has_value());
+          if (not python_error_message.has_value() == error_msg.has_value()) {
+            CAPTURE(python_error_msg_function);
+            CAPTURE(*python_error_message);
+            CAPTURE(*error_msg);
+            REQUIRE(python_error_message.has_value() == error_msg.has_value());
```

Comment:
I can see how this isn't obvious, and maybe you have a solution I haven't thought of. My struggle is that I want to capture the error messages (so they get printed) if the state is wrong, but I can't safely dereference an optional without first checking that it is valid (which I just noticed is not being done correctly anyway, sigh). Basically, I want to CAPTURE the error messages, but can only do so if they're valid. Maybe adding a lambda `get_error_msg_if_valid` and using that in the `CAPTURE`s would be better. I still need to guard the regex `CHECK` with either a REQUIRE or an if block

Reply:
Oh I see.. I to agree moving this into a lambda would help, if only to help "scope" the logic and add a few comments detailing the reason

Reply:
I've switched to `value_or`, but I don't like having a helper lambda for all of it (e.g. the CAPTUREs) because then the reported lines are not helpful (in the lambda, but you care about the calling context). Since this is already fairly complicated code, if someone needs to look at it to get context for the check failure, I'd like to make that as easy as possible